File: javav2/example_code/cognito/src/main/java/com/example/cognito/CreateUser.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.example.cognito;

// snippet-start:[cognito.java2.new_admin_user.main]
// snippet-start:[cognito.java2.new_admin_user.import]
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.cognitoidentityprovider.CognitoIdentityProviderClient;
import software.amazon.awssdk.services.cognitoidentityprovider.model.AdminCreateUserRequest;
import software.amazon.awssdk.services.cognitoidentityprovider.model.AdminCreateUserResponse;
import software.amazon.awssdk.services.cognitoidentityprovider.model.AttributeType;
import software.amazon.awssdk.services.cognitoidentityprovider.model.CognitoIdentityProviderException;
// snippet-end:[cognito.java2.new_admin_user.import]

/**
 * Before running this Java V2 code example, set up your development
 * environment, including your credentials.
 *
 * For more information, see the following documentation topic:
 *
 * https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/get-started.html
 */
public class CreateUser {
    public static void main(String[] args) {
        final String usage = """

                Usage:
                    <userPoolId> <userName> <email> <password>

                Where:
                    userPoolId - The Id value for the user pool where the user is created.
                    userName - The user name for the new user.
                    email - The email to use for verifying the user.
                    password - The password for this user.
                """;

        if (args.length != 4) {
            System.out.println(usage);
            System.exit(1);
        }

        String userPoolId = args[0];
        String userName = args[1];
        String email = args[2];
        String password = args[3];

        CognitoIdentityProviderClient cognitoClient = CognitoIdentityProviderClient.builder()
                .region(Region.US_EAST_1)
                .build();

        createNewUser(cognitoClient, userPoolId, userName, email, password);
        cognitoClient.close();
    }

    public static void createNewUser(CognitoIdentityProviderClient cognitoClient,
            String userPoolId,
            String name,
            String email,
            String password) {

        try {
            AttributeType userAttrs = AttributeType.builder()
                    .name("email")
                    .value(email)
                    .build();

            AdminCreateUserRequest userRequest = AdminCreateUserRequest.builder()
                    .userPoolId(userPoolId)
                    .username(name)
                    .temporaryPassword(password)
                    .userAttributes(userAttrs)
                    .messageAction("SUPPRESS")
                    .build();

            AdminCreateUserResponse response = cognitoClient.adminCreateUser(userRequest);
            System.out.println(
                    "User " + response.user().username() + "is created. Status: " + response.user().userStatus());

        } catch (CognitoIdentityProviderException e) {
            System.err.println(e.awsErrorDetails().errorMessage());
            System.exit(1);
        }
    }
}
// snippet-end:[cognito.java2.new_admin_user.main]

File: kotlin/services/cognito/src/main/kotlin/com/kotlin/cognito/CreateUser.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
package com.kotlin.cognito

// snippet-start:[cognito.kotlin.new_admin_user.import]
import aws.sdk.kotlin.services.cognitoidentityprovider.CognitoIdentityProviderClient
import aws.sdk.kotlin.services.cognitoidentityprovider.model.AdminCreateUserRequest
import aws.sdk.kotlin.services.cognitoidentityprovider.model.AttributeType
import kotlin.system.exitProcess
// snippet-end:[cognito.kotlin.new_admin_user.import]

/**
Before running this Kotlin code example, set up your development environment,
including your credentials.

For more information, see the following documentation topic:
https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/setup.html
 */

suspend fun main(args: Array<String>) {
    val usage = """
    Usage:
        <userPoolId> <userName> <email> <password>

    Where:
        userPoolId - The Id value for the user pool where the user will be created.
        userName - The user name.
        email - The email to use for verifying the account.
        password - The password to use (the characters that are allowed are uppercase, lowercase letters, numbers and at least one special character).
    """

    if (args.size != 4) {
        println(usage)
        exitProcess(0)
    }

    val userPoolId = args[0]
    val userName = args[1]
    val email = args[2]
    val password = args[3]
    createNewUser(userPoolId, userName, email, password)
}

// snippet-start:[cognito.kotlin.new_admin_user.main]
suspend fun createNewUser(
    userPoolId: String,
    name: String,
    email: String,
    password: String,
) {
    val attType =
        AttributeType {
            this.name = "email"
            value = email
        }

    val request =
        AdminCreateUserRequest {
            this.userPoolId = userPoolId
            username = name
            temporaryPassword = password
            userAttributes = listOf(attType)
        }

    CognitoIdentityProviderClient { region = "us-east-1" }.use { cognitoClient ->
        val response = cognitoClient.adminCreateUser(request)
        println("User ${response.user?.username} is created. Status is ${response.user?.userStatus}")
    }
}
// snippet-end:[cognito.kotlin.new_admin_user.main]
